pallets/tinlake-investor-pool/src/lib.rs

```diff
@@ -0,0 +1,1041 @@
+				.tranches
+				.iter_mut()
+				.enumerate()
+				.zip(solution.iter().copied())
+				.zip(execution.iter().copied())
+				.zip(&epoch.tranches)
+			{
+				let loc = TrancheLocator {
+					pool_id,
+					tranche_id: T::TrancheId::try_from(tranche_id)
+						.map_err(|_| Error::<T>::TrancheId)?,
+				};
+				Self::update_tranche_for_epoch(
+					loc,
+					pool.last_epoch_executed,
+					tranche,
+					solution,
+					execution,
+					epoch_tranche.price,
+				)?;
+			}
+
+			// Update the total/available reserve for the new total value of the pool
+			pool.total_reserve = pool
+				.total_reserve
+				.checked_add(&total_supply)
+				.and_then(|res| res.checked_sub(&total_redeem))
+				.ok_or(Error::<T>::Overflow)?;
+			pool.available_reserve = pool.total_reserve;
+
+			// Calculate the new fraction of the total pool value that each tranche contains
+			// This is based on the tranche values at time of epoch close.
+			let total_assets = pool
+				.total_reserve
+				.checked_add(&epoch.nav)
+				.ok_or(Error::<T>::Overflow)?;
+			let tranche_ratios: Vec<_> = execution
+				.iter()
+				.zip(&epoch.tranches)
+				.map(|((supply, redeem), tranche)| {
+					tranche
+						.value
+						.checked_add(supply)
+						.and_then(|value| value.checked_sub(redeem))
+						.map(|tranche_asset| {
+							Perquintill::from_rational(tranche_asset, total_assets)
+						})
+				})
+				.collect::<Option<Vec<Perquintill>>>()
+				.ok_or(Error::<T>::Overflow)?;
+
+			// Calculate the new total asset value for each tranche
+			// This uses the current state of the tranches, rather than the cached epoch-close-time values.
+			let mut total_assets = total_assets;
+			let tranche_assets = execution
+				.iter()
+				.zip(&mut pool.tranches)
+				.map(|((supply, redeem), tranche)| {
+					Self::update_tranche_debt(tranche)?;
+					tranche
+						.debt
+						.checked_add(&tranche.reserve)
+						.and_then(|value| value.checked_add(supply))
+						.and_then(|value| value.checked_sub(redeem))
+						.map(|value| {
+							if value > total_assets {
+								let assets = total_assets;
+								total_assets = Zero::zero();
+								assets
+							} else {
+								total_assets = total_assets.saturating_sub(value);
+								value
+							}
+						})
+				})
+				.collect::<Option<Vec<T::Balance>>>()
+				.ok_or(Error::<T>::Overflow)?;
+
+			// Rebalance tranches based on the new tranche asset values and ratios
+			let nav = T::NAV::nav(pool_id).ok_or(Error::<T>::NoNAV)?.0.into();
+			let mut remaining_nav = nav;
+			let mut remaining_reserve = pool.total_reserve;
+			let last_tranche = pool.tranches.len() - 1;
+			for (((tranche_id, tranche), ratio), value) in pool
+				.tranches
+				.iter_mut()
+				.enumerate()
+				.zip(&tranche_ratios)
+				.zip(&tranche_assets)
+			{
+				tranche.ratio = *ratio;
+				if tranche_id == last_tranche {
+					tranche.debt = remaining_nav;
+					tranche.reserve = remaining_reserve;
+				} else {
+					tranche.debt = ratio.mul_ceil(nav);
```

Comment:
shouldn;t this be the debtratio instead of tranch share of the pool ?

Reply:
We don't track separate debt and reserve ratios. We track a single pool ratio, and the goal of this step is to readjust the debt/reserve values of each tranche based on the new debt/reserve totals and the tranche's recomputed ratio from the epoch close.